### DISCUSSION TITLE
QQ 1.- What is executed first: Expect assertions or AfterResponse hook? 2.- How to fail test programatically from 'processor' js file?

### DISCUSSION THREAD

**Discussion by eliezercazares** [artilleryio/artillery#2296] [Artillery]
Hi all,
I have two questions.

1.- What is executed first: Expect assertions or AfterResponse hook?

2.- How to fail test programmatically from 'processor' js file?

Thanks a lot for your time and help

**Comment by hassy** [comment]
hey @eliezercazares 👋 

`expect` assertions are implemented with an `afterResponse` hook themselves internally. There's no strict ordering for `afterResponse` hooks - are you running into issues where you need a specific order?

There's no way to fail an entire test from custom JS code yet, but individual VUs can be failed by calling `next()` with an `Error` object - https://www.artillery.io/docs/reference/engines/http#function-signatures

**Comment by eliezercazares** [comment]
Hello Hassy!
Hope all is doing well there.

I was having issues getting an "afterResponse" variable validated with "Expect" in the YAML. the value was arriving Null. However I changed the logic to use async/sync as needed within the "Processor" file and now I am getting the value constantly. 
Thanks for the explanation on Expect and Hooks because it is really good to know!

In the other side, I am failing tests programmatically by using "Throw new error()" (vanilla JS) but I see what you mention about an Error Object.
I searched in the docs but couldn't find an example... Does this mean that I can just assign an error object to next()? such as 
```
next().error = "My error"
```
Sorry for my need for examples, I did try to find an example.

Thanks for your help, Artillery rocks!

**Reply by hassy** [reply]
> I searched in the docs but couldn't find an example... Does this mean that I can just assign an error object to next()? such as

You're right, we'll update the docs! I meant something like this:

```js
function myAfterResponseHook(req, res, userContext, events, next) {
  // do something with the response
  // ...
  // fail the VUs:
  return next(new Error('some problem with the response'));
}
```

This will do two things:

1. It will stop any remaining steps in the scenario for this VU from executing
2. The error will be logged as part of the metrics output from Artillery


> Thanks for your help, Artillery rocks!

Thanks! :D
